File: 프로그래머스/2/169198. 당구 연습/당구 연습.java
class Solution {
    public int[] solution(int m, int n, int startX, int startY, int[][] balls) {
        int[][] starts = new int[][] {
            {-startX, startY},
            {startX, 2*n -startY},
            {2*m - startX, startY},
            {startX, -startY},
        };
        
        int[] answer = new int[balls.length];
        
        for (int i = 0; i < balls.length; i++) {
            int minimum = 2000000;
            int[] calc = new int[]{0,1,2,3};
            
            if (startX == balls[i][0]) {
                if (startY > balls[i][1]) {
                    calc = new int[]{0,1,2};
                }
                else {
                    calc = new int[]{0,2,3};
                }
            }
            
            if (startY == balls[i][1]) {
                if (startX > balls[i][0]) {
                    calc = new int[]{1,2,3};
                }
                else {
                    calc = new int[]{0,1,3};
                }
            }
            
            for (int c : calc) {
                int s_x = starts[c][0];
                int s_y = starts[c][1];
                
                minimum = (int) Math.min(minimum, Math.pow(s_x - balls[i][0], 2) + Math.pow(s_y - balls[i][1], 2)); 
            }
            answer[i] = minimum;
        }
        return answer;
    }
}

File: 프로그래머스/2/169198. 당구 연습/당구 연습.py
# 8개 확인
# x = 0, x = m, y = 0, y = n
# (y,x) = (0,0), (n, 0), (n, m), (0, m)




def solution(m, n, startX, startY, balls):
        
    starts = [
            [-startX, startY], # 왼쪽 |
            [startX, 2 * n - startY], # 위 ㅡ
            [2 * m - startX, startY], # 오른쪽 |
            [startX, -startY], # 아래 ㅡ
        ]     
    answer =[]
    for ball in balls:
        minimum = 1e9
        calc = [0,1,2,3]
            
        if startX == ball[0]:
            if startY > ball[1]:
                calc = [0,1,2]
            else:
                calc = [0,2,3]
        
        if startY == ball[1]:
            if startX > ball[0]:
                calc = [1,2,3]
            else:
                calc = [0,1,3]
        
        for c in calc:
            start_x, start_y = starts[c]
            minimum = min(minimum, (start_x - ball[0]) ** 2 + (start_y - ball[1]) ** 2)
        
        answer.append(minimum)
    
    return answer
        
